### PR TITLE
Leverage TypeDB Cluster's database management coordination

### PR DESCRIPTION
**Pull request by lolski**
## What is the goal of this PR?

We have removed the workaround for TypeDB Cluster's database management operations. Prior to this PR, the client executes the operation on each server in the cluster.

In this PR, the client only goes to one server and let that server coordinate the operation to the other servers. This is made possible by leveraging the new database management coordination feature that has been implemented in TypeDB Cluster.

## What are the changes implemented in this PR?

- `create`, `get`, and `contains` calls go to any replica, but may optionally go to the primary replica.
- `delete` call must go to the primary replica

### CODE REVIEW

File: cluster/ClusterDatabase.java [hunk 1]
```diff
@@ -79,9 +81,20 @@ public String schema() {
...
+                return null;
+            }
+
+            @Override
+            Void rerun(Replica replica) {
+                run(replica);
+                return null;
+            }
+        };
+        failsafeTask.runPrimaryReplica();
```
Comment (lolski) [inline cluster/ClusterDatabase.java:97]: Perform delete operation to the primary replica.

File: cluster/ClusterDatabaseManager.java [hunk 2]
```diff
@@ -100,4 +97,26 @@ public void create(String name) {
...
+            RESULT run(ClusterDatabase.Replica replica) {
+                return task.apply(client.stub(replica.address()), client.coreClient(replica.address()).databases());
+            }
+
+            @Override
+            RESULT rerun(ClusterDatabase.Replica replica) {
+                return run(replica);
+            }
+        };
+        try {
```
Comment (lolski) [inline cluster/ClusterDatabaseManager.java:114]: `failsafeTask` performs `task` to any replica, and if that replica isn't the primary replica, will retry to the primary replica.

This is unintuitive and will be fixed in the upcoming PRs. The fix will require some changes on the server side.

File: cluster/ClusterDatabase.java [hunk 3]
```diff
@@ -41,11 +41,13 @@
     private static final Logger LOG = LoggerFactory.getLogger(ClusterDatabase.class);
     private final String name;
     private final Map<String, CoreDatabase> databases;
+    private ClusterClient client;
     private final ClusterDatabaseManager databaseMgr;
     private final Set<Replica> replicas;
 
-    private ClusterDatabase(String database, ClusterDatabaseManager clusterDatabaseMgr) {
+    private ClusterDatabase(String database, ClusterClient client, ClusterDatabaseManager clusterDatabaseMgr) {
```
Comment (alexjpwalker) [inline cluster/ClusterDatabase.java:48]: If we're passing in the whole Client, might as well grab `databases` from it:
```suggestion
    private ClusterDatabase(String database, ClusterClient client) {
```

```java
this.databaseMgr = client.databases();
```

File: cluster/ClusterDatabase.java [hunk 4]
```diff
@@ -55,10 +57,10 @@ private ClusterDatabase(String database, ClusterDatabaseManager clusterDatabaseM
         }
     }
 
-    static ClusterDatabase of(ClusterDatabaseProto.ClusterDatabase protoDB, ClusterDatabaseManager clusterDatabaseMgr) {
+    static ClusterDatabase of(ClusterDatabaseProto.ClusterDatabase protoDB, ClusterClient client, ClusterDatabaseManager clusterDatabaseMgr) {
```
Comment (alexjpwalker) [inline cluster/ClusterDatabase.java:60]: see above

Comment (lolski) [reply cluster/ClusterDatabase.java]: Fixed in dc11f37cc5faade1de8a9f377923eb87269439f3.

**Review by lolski — COMMENTED** [review]
@alexjpwalker Don't mind the merge conflict. It is expected and I'll fix all of it once Haikal merged his renaming PRs.

**Comment by lolski** [comment]
Superseded by https://github.com/vaticle/typedb-client-java/pull/328

**Pull request closed without merging**